[Clang][OpenMP] Support for dispatch construct (Sema & Codegen) support

SunilKuravinakop opened the pull request:
Support for dispatch construct (Sema & Codegen) support. Support for clauses: depend, novariants & nocontext.

Commit:
Support for dispatch construct (Sema & Codegen) support. Support for clauses depend, novariants & nocontext.

Commit:
Taking care of feedback comments from Alexey Bataev.
1) Changing comments from // to ///
2) Using STLExtras like llvm::is_contained().
3) Insteading of combining function comments for 2 functions provided
   them separately before the function definitions.

Commit:
Changed the comments for replaceWithNewTraitsOrDirectCall().

Commit:
Removing an unnecessary bracket in if statement in ActOnOpenMPCall().

Commit:
Removing unnecessary brackets around if statement.
Instead of using nullptr for initilizalition StmtResult using an
uninitialized StmtResult.
Unnecessary StmtResult & llvm::ArrayRef variables.

Commit:
1) Claiming support for "dispatch" construct.
2) Changed comments in transformDispatchDirective() in SemaOpenMP.cpp
3) Adding checks in clang/test/OpenMP/dispatch_messages.cpp for
        a) depend clause
        b) nocontext, novariants & depend clauses occuring in the same line of dispatch construct.
4) Removing debugging statements (kept under #if 0 and #endif) in OMPContext.h.

Commit:
Spacing problems for claiming "dispatch construct".

Commit:
Adding support for dispatch construct in Release Notes.

Commit:
re-wording message in warn_omp_dispatch_clause_novariants_nocontext.

Commit:
Changes in expected-warning message for "#pragma omp dispatch
nocontext(c1) novariants(c2) depend(inout:a)".

Commit:
Removing the word CodeGen from comments, which can lead to confusion.

Commit:
Moving the helper functions to CodeGen by using AnnotateAttr in Sema.

Commit:
Changing the commit message from:
  'nocontext' clause is ignored, only 'novariants' clause is applied
to
  first occuring clause among 'nocontext' or 'novariants' is applied
This is being done because of usage of llvm::any_of & llvm::is_contained instead of if statements.

Commit:
All clang/test/OpenMP/dispatch_*.cpp tests pass with the changes in this commit.

Commit:
Merge branch 'main' into dispatch_depend

Commit:
Handling:
1) Templates
2) Constant values in the conditions for novariants & nocontext.

Commit:
Undoing changes to handle templates.

Commit:
Merge branch 'main' into dispatch_depend

Commit:
Merge branch 'main' into dispatch_depend

alexey-bataev commented:
The codegen for each(virtual) directive should live on codegen. But(!) Sema should emit separate (specific) CapturedStmts for each such virtual region. Check, how the combined directives work. Sema defines multiple CapturedStmt for each combined directive, and then codegen emits the code for each(virtual) standalone directive, using CapturedStmt for this particular (virtual) standalone directive. The list of captures of each sub-directives might be different, so Sema should provide different CapturedStmt for each emitted sub-directive.

Commit:
Separating virtual regions of taskwait directive and dispatch directive
in Sema.

Commit:
Merge branch 'main' into dispatch_depend

Commit:
Merge branch 'main' into dispatch_depend

Commit:
Merge branch 'main' into dispatch_depend

SunilKuravinakop commented:
> The codegen for each(virtual) directive should live on codegen. But(!) Sema should emit separate (specific) CapturedStmts for each such virtual region. Check, how the combined directives work. Sema defines multiple CapturedStmt for each combined directive, and then codegen emits the code for each(virtual) standalone directive, using CapturedStmt for this particular (virtual) standalone directive. The list of captures of each sub-directives might be different, so Sema should provide different CapturedStmt for each emitted sub-directive.

With the present checkin the -ast-dump looks as follows:
```
     |-CompoundStmt 0x17f4a10 <<invalid sloc>>
     |  |-OMPTaskwaitDirective 0x1838208 <line:29:5, col:54> openmp_standalone_directive
     |  |  `-OMPDependClause 0x1837dc0 <col:40, col:53>
     |  |   `-DeclRefExpr 0x1837d88 <col:52> 'int' lvalue Var 0x1837890 'x' 'int'
     |  `-OMPDispatchDirective 0x17f49c0 <col:5, col:54>
     |    |-OMPNocontextClause 0x1837d60 <col:26, col:38>
     |    |  `-ImplicitCastExpr 0x1837d30 <col:36> 'int' <LValueToRValue>
     |    |    `-DeclRefExpr 0x1837d10 <col:36> 'int' lvalue OMPCapturedExpr 0x1837c90 '.capture_expr.' 'int'
     |    |-OMPDependClause 0x1837dc0 <col:40, col:53>
     |    |       `-DeclRefExpr 0x1837d88 <col:52> 'int' lvalue Var 0x1837890 'x' 'int'
     |    `-CapturedStmt 0x18381c8 <line:30:5, col:11>
     |        |-CapturedDecl 0x1837ef8 <<invalid sloc>> <invalid sloc>

```
Since, there is no need for a CapturedStmt below OMPTaskDirective none has been added.  Can you please review the code.

alexey-bataev commented:
> > The codegen for each(virtual) directive should live on codegen. But(!) Sema should emit separate (specific) CapturedStmts for each such virtual region. Check, how the combined directives work. Sema defines multiple CapturedStmt for each combined directive, and then codegen emits the code for each(virtual) standalone directive, using CapturedStmt for this particular (virtual) standalone directive. The list of captures of each sub-directives might be different, so Sema should provide different CapturedStmt for each emitted sub-directive.
> 
> With the present checkin the -ast-dump looks as follows:
> 
> ```
>      |-CompoundStmt 0x17f4a10 <<invalid sloc>>
>      |  |-OMPTaskwaitDirective 0x1838208 <line:29:5, col:54> openmp_standalone_directive
>      |  |  `-OMPDependClause 0x1837dc0 <col:40, col:53>
>      |  |   `-DeclRefExpr 0x1837d88 <col:52> 'int' lvalue Var 0x1837890 'x' 'int'
>      |  `-OMPDispatchDirective 0x17f49c0 <col:5, col:54>
>      |    |-OMPNocontextClause 0x1837d60 <col:26, col:38>
>      |    |  `-ImplicitCastExpr 0x1837d30 <col:36> 'int' <LValueToRValue>
>      |    |    `-DeclRefExpr 0x1837d10 <col:36> 'int' lvalue OMPCapturedExpr 0x1837c90 '.capture_expr.' 'int'
>      |    |-OMPDependClause 0x1837dc0 <col:40, col:53>
>      |    |       `-DeclRefExpr 0x1837d88 <col:52> 'int' lvalue Var 0x1837890 'x' 'int'
>      |    `-CapturedStmt 0x18381c8 <line:30:5, col:11>
>      |        |-CapturedDecl 0x1837ef8 <<invalid sloc>> <invalid sloc>
> ```
> 
> Since, there is no need for a CapturedStmt below OMPTaskDirective none has been added. This is being processed properly and -emit-llvm produces the correct output. Can you please review the code?

There should not be taskwait directive in AST, it should be emitted in codegen. Only corresponding CapturedStmt should be built in Sema

SunilKuravinakop commented:
There is only 1 "dispatch" directive. How can we have multiple virtual regions? If we emit "taskwait" directive in AST then we will have multiple virtual regions. 
I have looked at combined directives but, in them we have multiple directives.
```
#pragma omp parallel
  #pragma omp for
```
becomes:
```
    `-OMPParallelDirective 0x1311798 <line:10:3, col:24>
      `-CapturedStmt 0x1311770 <line:11:6, col:21>
        `-CapturedDecl 0x12a5770 <<invalid sloc>> <invalid sloc> nothrow
          |-OMPForDirective 0x1311668 <col:6, col:21>
          | `-CapturedStmt 0x12a5f10 <line:12:6, line:4:11>

```
So, I do not understand what you have said. Can you please clarify? Moving back emitting of taskwait directive back into CodeGen is not much of a problem.

dreachem commented:
@alexey-bataev For something like this:

```c++
void f()
{
    #pragma omp dispatch nowait depend(inout:x)
    g();
}
```

The AST dump for that function is as follows, with @SunilKuravinakop's patch:

```
`-FunctionDecl 0xed2dac8 <line:7:1, line:11:1> line:7:6 f 'void ()'
  `-CompoundStmt 0xed2df28 <line:8:1, line:11:1>
    `-CompoundStmt 0xed2df08 <<invalid sloc>>
      |-OMPTaskwaitDirective 0xed2ded0 <line:9:5, col:48> openmp_standalone_directive
      | `-OMPDependClause 0xed2dbe8 <col:33, col:47>
      |   `-DeclRefExpr 0xed2db80 <col:46> 'int' lvalue Var 0xed2da38 'x' 'int'
      `-CapturedStmt 0xed2dea8 <line:10:5, col:7>
        `-CapturedDecl 0xed2dd18 <<invalid sloc>> <invalid sloc>
          |-CallExpr 0xed2de88 <col:5, col:7> 'void'
          | `-ImplicitCastExpr 0xed2de70 <col:5> 'void (*)()' <FunctionToPointerDecay>
          |   `-DeclRefExpr 0xed2de20 <col:5> 'void ()' Function 0xed2d7e0 'g' 'void ()'
          `-ImplicitParamDecl 0xed2ddb0 <line:9:5> col:5 implicit __context 'struct (unnamed at dispatch.c:9:5) *const restrict'
```

Should the AST instead look something like this?

```
`-FunctionDecl 0xf804ac8 <line:7:1, line:11:1> line:7:6 f 'void ()'
  `-CompoundStmt 0xf804ea8 <line:8:1, line:11:1>
    `-OMPDispatchDirective 0xf804e60 <line:9:5, col:57>
      |-OMPNowaitClause 0xf804b70 <col:26, col:57>'
      |-OMPDependClause ...
      `-CapturedStmt 0xf804e08 <line:10:5, col:7>
        `-CapturedDecl 0xf804c78 <<invalid sloc>> <invalid sloc>
          |-CallExpr 0xf804de8 <col:5, col:7> 'void'
          | `-ImplicitCastExpr 0xf804dd0 <col:5> 'void (*)()' <FunctionToPointerDecay>
          |   `-DeclRefExpr 0xf804d80 <col:5> 'void ()' Function 0xf8047e0 'g' 'void ()'
          `-ImplicitParamDecl 0xf804d10 <line:9:5> col:5 implicit __context 'struct (unnamed at dispatch.c:9:5) *const restrict'
```

and then handle the implicit taskwait in codegen? I am also confused how an additional CapturedStmt to represent the taskwait should be implemented in Sema.

alexey-bataev commented:
> Should the AST instead look something like this?

Not quite, there should be outer CapturedStmt for (virtual!) taskwait directive.
E.g.
```
void foo() {
#pragma omp target nowait
  ;
}

```
results in
```
`-FunctionDecl 0x55776d5e6e20 <repro.cpp:1:1, line:4:1> line:1:6 foo 'void ()'
  `-CompoundStmt 0x55776d5e77f0 <col:12, line:4:1>
    `-OMPTargetDirective 0x55776d5e77b0 <line:2:1, col:26>
      |-OMPNowaitClause 0x55776d5e6f10 <col:20, col:26>
      `-CapturedStmt 0x55776d5e7788 <line:3:3>
        `-CapturedDecl 0x55776d5e7170 <<invalid sloc>> <invalid sloc> nothrow
          |-CapturedStmt 0x55776d5e7760 <col:3>
          | `-CapturedDecl 0x55776d5e7660 <<invalid sloc>> <invalid sloc> nothrow
          |   |-NullStmt 0x55776d5e7758 <col:3>
          |   `-ImplicitParamDecl 0x55776d5e76f0 <line:2:1> col:1 implicit __context '(unnamed struct at repro.cpp:2:1) *const __restrict'
          |-AlwaysInlineAttr 0x55776d5e7498 <<invalid sloc>> Implicit __forceinline
          |-ImplicitParamDecl 0x55776d5e71f8 <col:1> col:1 implicit .global_tid. 'const int'
          |-ImplicitParamDecl 0x55776d5e7260 <col:1> col:1 implicit .part_id. 'const int *const __restrict'
          |-ImplicitParamDecl 0x55776d5e72c8 <col:1> col:1 implicit .privates. 'void *const __restrict'
          |-ImplicitParamDecl 0x55776d5e7330 <col:1> col:1 implicit .copy_fn. 'void (*const __restrict)(void *const __restrict, ...)'
          |-ImplicitParamDecl 0x55776d5e7398 <col:1> col:1 implicit .task_t. 'void *const'
          |-ImplicitParamDecl 0x55776d5e7430 <col:1> col:1 implicit __context '(unnamed struct at repro.cpp:2:1) *const __restrict'
          |-CXXRecordDecl 0x55776d5e74f0 <col:1> col:1 implicit struct definition
          | |-DefinitionData pass_in_registers empty aggregate standard_layout trivially_copyable pod trivial literal has_constexpr_non_copy_move_ctor can_const_default_init
          | | |-DefaultConstructor exists trivial constexpr needs_implicit defaulted_is_constexpr
          | | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
          | | |-MoveConstructor exists simple trivial needs_implicit
          | | |-CopyAssignment simple trivial has_const_param needs_implicit implicit_has_const_param
          | | |-MoveAssignment exists simple trivial needs_implicit
          | | `-Destructor simple irrelevant trivial needs_implicit
          | `-CapturedRecordAttr 0x55776d5e75a0 <<invalid sloc>> Implicit
          `-CapturedDecl 0x55776d5e7660 <<invalid sloc>> <invalid sloc> nothrow
            |-NullStmt 0x55776d5e7758 <line:3:3>
            `-ImplicitParamDecl 0x55776d5e76f0 <line:2:1> col:1 implicit __context '(unnamed struct at repro.cpp:2:1) *const __restrict'
```
You can see 2 CpaturedStmt here - one for outer task region (outermost one) and one for the actual target region (the innermost one). You need to do something similar for dispatch

Commit:
Moving handling of dispatch depend into CodeGen.

Commit:
Changes to be committed:
 	modified:   clang/lib/CodeGen/CGStmtOpenMP.cpp

Commit:
-ast-dump is working with addition of extra virtual region. CodeGen is
still in progress.

Commit:
modified:   clang/lib/Sema/SemaOpenMP.cpp

SunilKuravinakop commented:
> You can see 2 CpaturedStmt here - one for outer task region (outermost one) and one for the actual target region (the innermost one). You need to do something similar for dispatch

I have done a commit where -ast-dump works now. CodeGen is still failing for `EmitIfElse` . Please give me a feedback if I am doing it correctly.

Commit:
Formatting problems.
 	modified:   clang/lib/Sema/SemaOpenMP.cpp

Commit:
CodeGen works with this change.
 	modified:   clang/lib/CodeGen/CGStmtOpenMP.cpp

Commit:
All tests pass with this change.

alexey-bataev commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -537,6 +537,10 @@ Python Binding Changes
 
 OpenMP Support
 --------------
+- Added support for 'omp assume' directive.
+- Added support for 'omp scope' directive.
+- Added support for allocator-modifier in 'allocate' clause.
```
Unrelated to this patch

SunilKuravinakop replied:
This is the result of pulling in changes from github. I will try to correct it.

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -4528,6 +4528,115 @@ void CodeGenFunction::EmitOMPMasterDirective(const OMPMasterDirective &S) {
   emitMaster(*this, S);
 }
 
+static Expr *getInitialExprFromCapturedExpr(Expr *Cond) {
```
Why do you need it? CapturedExpr should be emitted as is

SunilKuravinakop replied:
It is my mistake in naming the function as getInitialExprFromCapturedExpr(). It should have been getCapturedExprFromImplicitCastExpr( ). The function extracts the OMPCapturedExpr from within the clause.  
clause->getCondition() does not give me the OMPCapturedExpr. 
```
(gdb) NoContextC->getCondition()->dump()
ImplicitCastExpr 0x678640 'int' <LValueToRValue>
  `-DeclRefExpr 0x678620 'int' lvalue OMPCapturedExpr 0x6785a0 '.capture_expr.' 'int'
```
`NoContextC `is of type `OMPNocontextClause *`.
If there is a simpler way to get the OMPCapturedExpr please tell me.

At present, I will rename the function name.

SunilKuravinakop replied:
While generating the If-else statement in EmitIfElse(), I am using 
`llvm::Value *CondValue = CGF->EvaluateExprAsBool(Condition);`
and
`CGF->Builder.CreateCondBr(CondValue, ThenBlock, ElseBlock);
`
The `EvaluateExprAsBool `needs a `CapturedExpr.` Using `NoContextC->getCondition()` results in a core dump.

alexey-bataev replied:
It means, the context (capturedstmt) does not know about this capturedexpr, which means it is captured in the wrong context. You're not fixing the problem, you're overcoming it with this hack. Need to capture the expression in the correct context and emit in the correct context

SunilKuravinakop replied:
I use S.clauses() to get the list of all the clauses (type OMPClause) present in the dispatch directive. The example of NoContextC of type OMPClause contains ImplicitCastExpr. An OMPClause cannot be changed to contain capturedExpr. I think I am not able to understand where do you want me to change it. Can you please clarify?

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -4528,6 +4528,115 @@ void CodeGenFunction::EmitOMPMasterDirective(const OMPMasterDirective &S) {
+    CapturedStmtInfo = &CapStmtInfo;
+
+  if (!Clauses.empty()) {
+    if (S.hasClausesOfKind<OMPDependClause>())
+      EmitOMPDispatchToTaskwaitDirective(S);
```
If the task region is not needed, use `CGM.getOpenMPRuntime().emitInlinedDirective` to emit inner regions

SunilKuravinakop replied:
For `#pragma omp dispatch novariant(c2)`  a task region is not needed. But, this results in :
```
If (c2) {
   foo();
 } else {
    foo_variant();
}
```
Here how can I use `CGM.getOpenMPRuntime().emitInlinedDirective`  to generate inlined if-else or the calls within it?

alexey-bataev replied:
Try to insert something like `OMPLexicalScope Scope(CGF, S, OMPD_unknown);` before emitting if-based logic

SunilKuravinakop replied:
I have inserted 
`OMPLexicalScope Scope(*this, S, OMPD_dispatch);` instead of `OMPLexicalScope Scope(CGF, S, OMPD_unknown);`

SunilKuravinakop replied:
If I avoid getCapturedExprFromImplicitCastExpr then I will have to use :
`emitIfElse(this, NoVariantsC->getCondition(), AssociatedStmt).`
This will result in using assertion while generating the `if ` statement because the condition is an ImplicitCastExpr and not CapturedExpr. I have verified this.

alexey-bataev replied:
Again, it means that you're not correctly handling the context. Check how it is implemented in other constructs.

SunilKuravinakop replied:
Using, EmitIfStmt() is probably one of the ways by creating instance of IfStmt and call EmitStmt(IfStmt). If there is a better way than this please inform me. 
The current implementation that I have in emitIfElse() was optimised:

1. to use the AssociatedStmt in generating ElseStmt 
2. then replace the call in AssociatedStmt (e.g. from foo_variant() to foo())
3. Again use the AssociatedStmt in generating ThenStmt

But, if I use the EmitIfStmt() or any other function then this is not possible.

alexey-bataev reviewed (commented):
Refer to target region codegen

Commit:
1) Changing the name of the function from
getInitialExprFromCapturedExpr()
    to
getCapturedExprFromImplicitCastExpr()

2) Removing extra additions from clang/docs/ReleaseNotes.rst.

Commit:
Merge branch 'main' into dispatch_depend

Commit:
Merge branch 'main' into dispatch_depend

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -4529,6 +4529,117 @@ void CodeGenFunction::EmitOMPMasterDirective(const OMPMasterDirective &S) {
+  }
+}
+
+static void EmitIfElse(CodeGenFunction *CGF, Expr *Condition,
```
```suggestion
static void emitIfElse(CodeGenFunction *CGF, Expr *Condition,
```

Commit:
Added OMPLexicalScope before emitting if-else statement.

Commit:
Changing LexicalScope to OMPD_dispatch.

SunilKuravinakop commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -11899,6 +11899,9 @@ def err_omp_clause_requires_dispatch_construct : Error<
   "'%0' clause requires 'dispatch' context selector">;
 def err_omp_append_args_with_varargs : Error<
   "'append_args' is not allowed with varargs functions">;
+def warn_omp_dispatch_clause_novariants_nocontext : Warning<
```
This message is to indicate the user that if novariants and nocontext are specified then novariants is taken into account. This warning is to inform the user and not an error.

alexey-bataev replied:
Why it is not an error?

SunilKuravinakop replied:
I do not need to stop execution if novariants and nocontext occur together.  In a dispatch directive like  `#pragma omp dispatch nocontext(c1) novariants(c2)` then I do not find a way to generate 
```
if(condition)
    foo();
else
   foo_variant

```
For the "condition" in `if `statement I do not know a way to use c1 & c2 together. Hence I preferred to indicate this to user and use `novariants(c2)`. I do not want to stop the compilation with an error message because the spec does not indicate what to do in this case.

SunilKuravinakop replied:
In case of :
```
#pragma omp declare variant(foo_variant2) match(user={condition(1)})
#pragma omp declare variant(foo_variant) match(construct={dispatch})
void foo(int x) {
}

#pragma omp dispatch nocontext(c1)
```
the code generated in the if-else is:
```
if(c1) {
   foo_variant2(x)
} else {
  foo_variant(x)
}
```
whereas for `#pragma omp dispatch novariantc2)` it is 
```
if(c2) {
   foo(x)
} else {
  foo_variant(x)
}
```
There is a difference in the code generate at `if` . Hence I cannot use `c1 || c2`.

alexey-bataev replied:
I see this in the standard:
If do-not-use-variant evaluates to true, no function variant is selected for the target-call of the
13 dispatch region associated with the novariants clause even if one would be selected
14 normally.

Doe it mean that novariant has higher priority than nocontext?

SunilKuravinakop replied:
This relates only to the novariants clause only. However, I can extrapolate the standard to mean whatever I have implemented. It will be good to provide a warning. 
If you feel that warning is not needed then I can remove it but, the user should not be left wondering why nocontext was not considered.

alexey-bataev replied:
@dreachem thoughts?

dreachem replied:
The effect of the **nocontext** clause is to not add the **dispatch** construct to the construct trait set. It will only matter once we support **dispatch** in a construct trait selector (support for this is not required in 5.2, but is required in 6.0).

Assuming we support **dispatch** in a construct trait selector, then this is what the various combinations would mean:

```
novariants(0), nocontext(0): Allow selective variant substitution according to context match
novariants(0), nocontext(1): Allow selective variant substitution according to context match (NO MATCH
                             for selectors that require **dispatch** trait)
novariants(1), nocontext(0): Always call the base function (shuts off any variant substitution for the call)
novariants(1), nocontext(1): Always call the base function (shuts off any variant substitution for the call)
```

In summary, **novariants** takes precedence and **nocontext** can be ignored ONLY if the _do-not-use-variant_ argument evaluates to true.

SunilKuravinakop replied:
Alexey,
    I have added support for clauses novariants and nocontext occuring on the same dispatch directive.
    Can you please review the code?
--Sunil

Commit:
Removed unused statements.

Commit:
Replacing getCapturedExprFromImplicitCastExpr() with
CGF->EmitBranchOnBoolExpr().

Commit:
Adding support for clauses novariants and nocontext occuring on the same
dispatch directive.

Commit:
Rolling back changes to make template related changes in this patch.
  Changes to be committed:
 	modified:   clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

SunilKuravinakop commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -423,8 +423,8 @@ static void instantiateOMPDeclareVariantAttr(
 
-  auto &&SubstExpr = [FD, ThisContext, &S, &TemplateArgs](Expr *E) {
-    if (auto *DRE = dyn_cast<DeclRefExpr>(E->IgnoreParenImpCasts()))
+  auto &&SubstExpr = [FD, ThisContext, &S, &TemplateArgs, New](Expr *E) {
```
Sorry, this was not deleted properly. I was attempting to include another change but, I subsequently rolled it back for this patch. Thanks for pointing it out.

SunilKuravinakop commented:
> Hmm, this PR is much shorter than it used to be.

The initial PR where you had given your comments was [117904](https://github.com/llvm/llvm-project/pull/117904) . I had made a push to wrong branch and as such 117904 was closed. I Could not re-open it.  I ended up creating a new PR but, this has all the changes. I have included a pointer to 117904 in the beginning of this PR.

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -4529,6 +4529,191 @@ void CodeGenFunction::EmitOMPMasterDirective(const OMPMasterDirective &S) {
+//     }
+//   }
+//
+static void emitIfElse(CodeGenFunction *CGF, Stmt *AssociatedStmt,
```
I rather doubt it will correctly with constructors/destructors

alexey-bataev replied:
Try to add tests with classes, which evaluate to boolean, and create immediate class instance and see, that the destructors are called correctly

dreachem replied:
@alexey-bataev Both @SunilKuravinakop and I are unclear what test you're asking for here.

When you say "test with classes", do you mean that the called function (and its variant) should have a local declaration of a variable of class type? Such as:

```c++
void f_v()
{
   MyType2 v; // MyType2 constructor called on v
   ...
   // MyType2 destructor called on v

}

#pragma omp declare variant(f_v) match(construct={dispatch})
void f()
{
   MyType v; // MyType constructor called on v
   ...
   // MyType destructor called on v
}

int main()
{
   #pragma omp dispatch
   f();
}
```

But then, what do you mean by "which evaluate to boolean"?

Can you provide a quick sketch of what this test should be doing?

alexey-bataev replied:
No, I thought about the condition, which involves instantiation of the class, to check that the codegen handles correctly constructors/destructors calls (that's always a problem in C++). Is it possible to create a condition, that includes class, something like:
```
class AAA {
...
};
#pragma omp dispatch if(AAA()) <---- the constructor and then destructor is called here
...
```

dreachem replied:
Thanks for the clarification. So something like this?

```c++
#include <iostream>                                                                                                                                                                                                                                                          
int main()
{
    class IsPositive {
public:
    IsPositive(int x) : _x(x) {
        std::cout << "constructing IsPositive(" << x << ") object\n";
    }
    ~IsPositive() {
        std::cout << "destructing IsPositive(" << _x << ") object\n";
    }
    operator bool() const { return _x > 0; }
private:
        int _x;
    };

  // case 1
#pragma omp parallel if(IsPositive(10))
    {
        std::cout << "[1] in parallel\n";
    }

  // case 2
    auto cond1 = IsPositive(5);
#pragma omp parallel if(cond1)
    {
        std::cout << "[2] in parallel\n";
    }

  // case 3
#pragma omp parallel if(IsPositive(-10))
    {
        std::cout << "[3] in parallel\n";
    }

  // case 4
    auto cond2 = IsPositive(-5);
#pragma omp parallel if(cond2)
    {
        std::cout << "[4] in parallel\n";
    }

  
```

The output for the above with `OMP_NUM_THREADS=2` is:

```
constructing IsPositive(10) object
[1] in parallel
[1] in parallel
destructing IsPositive(10) object
constructing IsPositive(5) object
[2] in parallel
[2] in parallel
constructing IsPositive(-10) object
[3] in parallel
destructing IsPositive(-10) object
constructing IsPositive(-5) object
[4] in parallel
destructing IsPositive(-5) object
destructing IsPositive(5) object
```

So, the destructor for the instantiated class object used in the `if` condition is called upon completing each `parallel` region. On the other hand, if I move the condition expression outside the directive (as for cases 2 and 4), then the destructor for the instantiated class is called at the end of the function.

Does this match the expected behavior?

alexey-bataev replied:
Yes, something like this but for dispatch. We need to handle classes construction/deconstruction correctly, I just want to be sure that we don't miss anything here and avoid bugs, that will require significant efforts to fix/rework later.

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -4529,6 +4529,191 @@ void CodeGenFunction::EmitOMPMasterDirective(const OMPMasterDirective &S) {
+  Stmt *CallExprStmt = CDecl->getBody();
+
+  if (BinaryOperator *BinaryCopyOpr = dyn_cast<BinaryOperator>(CallExprStmt)) {
+    CurrentCallExpr = BinaryCopyOpr->getRHS();
+    BinaryCopyOpr->setRHS(NewExpr);
+  } else {
+    CurrentCallExpr = dyn_cast<Expr>(CallExprStmt);
+    CDecl->setBody(NewExpr);
+  }
+
+  return CurrentCallExpr;
+}
+
+static Expr *transformCallInStmt(Stmt *StmtP, bool NoContext = false) {
+  Expr *CurrentExpr = nullptr;
+  if (auto *CptStmt = dyn_cast<CapturedStmt>(StmtP)) {
+    CapturedDecl *CDecl = CptStmt->getCapturedDecl();
+
+    CallExpr *NewCallExpr = nullptr;
+    for (const auto *attr : CDecl->attrs()) {
+      if (NoContext) {
+        if (const auto *annotateAttr =
+                llvm::dyn_cast<clang::AnnotateAttr>(attr);
+            annotateAttr && annotateAttr->getAnnotation() == "NoContextAttr") {
+          NewCallExpr = llvm::dyn_cast<CallExpr>(*annotateAttr->args_begin());
+        }
+      } else {
+        if (const auto *annotateAttr =
+                llvm::dyn_cast<clang::AnnotateAttr>(attr);
+            annotateAttr && annotateAttr->getAnnotation() == "NoVariantsAttr") {
+          NewCallExpr = llvm::dyn_cast<CallExpr>(*annotateAttr->args_begin());
+        }
+      }
+    }
+
+    CurrentExpr = replaceWithNewTraitsOrDirectCall(CDecl, NewCallExpr);
+  }
+  return CurrentExpr;
+}
+
```
I still think these functions should not be needed, if something is required, it should be build in Sema and stored in AST. If you need to replace some AST values by some LVM IR values, use OpaqueValue nodes, which can be replaced in codegen by special RAIIs.

SunilKuravinakop replied:
Can you please clarify your comment based on my following points?
1) There is a pre-computed call trait stored in AnnotateAttr. I am using those AnnotateAttrs in `transformCallInStmt`. In `replaceWithNewTraitsOrDirectCall` I am trying to set the call based on BinaryExpr under the dispatch directive.  
2) If you want me to pre-compute and store it in Sema I will have to clone the CDecl and store it. Earlier, in the helper functions (in Sema), I created new AST attributes, but according to your suggestion now I will have to store it in AnnotateAttr. 
3) An OpaqueValueExpr node copies the pointer of the SourceExpr. This does not help in noting the alternative call trait.

alexey-bataev replied:
The problem with these transformations is that sometimes they do not work as expected and miss some cases. Especially it happens with C++, which has lots of idioms like classes (especially inheritance), templates, implicit conversions (functions), etc. All these tree traversal techniques tend to be not fully correct in many cases. That;s why the preferred way to build helper expressions is doing it in Sema, because Sema knows how to handle all these cases correctly, but codegen does not. That's why I'm asking about compatibility with C++, it has lots of corner cases, and it would be good somehow to handle them in this patch rather than fixing it later

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -4529,6 +4529,191 @@ void CodeGenFunction::EmitOMPMasterDirective(const OMPMasterDirective &S) {
+    CGF->EmitBranchOnBoolExpr(Condition_NoVariants, ThenNoVariantsBlock,
+                              ElseNoVariantsBlock, 0);
+
+  } else if (Condition_NoVariants)
+    CGF->EmitBranchOnBoolExpr(Condition_NoVariants, ThenBlock, ElseBlock, 0);
+  else
+    CGF->EmitBranchOnBoolExpr(Condition_NoContext, ThenBlock, ElseBlock, 0);
```
```suggestion
  } else if (Condition_NoVariants) {
    CGF->EmitBranchOnBoolExpr(Condition_NoVariants, ThenBlock, ElseBlock, 0);
  } else {
    CGF->EmitBranchOnBoolExpr(Condition_NoContext, ThenBlock, ElseBlock, 0);
  }
```

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -4529,6 +4529,191 @@ void CodeGenFunction::EmitOMPMasterDirective(const OMPMasterDirective &S) {
+    CGF->EmitStmt(ThenNoContextStmt);
+    CGF->Builder.CreateBr(MergeBlock);
+
+  } else if (Condition_NoVariants) {
+    // Emit the NoVariants (then) block.
+    CGF->EmitBlock(ThenBlock);
+    Stmt *ThenStmt = AssociatedStmt;
+    ElseCall = transformCallInStmt(ThenStmt, false);
+    CGF->EmitStmt(ThenStmt);
+    CGF->Builder.CreateBr(MergeBlock);
+
+  } else if (Condition_NoContext) {
+    // Emit the NoContext (then) block.
+    CGF->EmitBlock(ThenBlock);
+    Stmt *ThenStmt = AssociatedStmt;
+    ElseCall = transformCallInStmt(ThenStmt, true);
+    CGF->EmitStmt(ThenStmt);
+    CGF->Builder.CreateBr(MergeBlock);
+  }
```
Same code, which differs only by boolean flag

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -4529,6 +4529,191 @@ void CodeGenFunction::EmitOMPMasterDirective(const OMPMasterDirective &S) {
+  ArrayRef<OMPClause *> Clauses = S.clauses();
+
+  Stmt *AssociatedStmt = const_cast<Stmt *>(S.getAssociatedStmt());
+  if (auto *AssocStmt = dyn_cast<CapturedStmt>(AssociatedStmt))
+    if (auto *InnerCapturedStmt =
+            dyn_cast<CapturedStmt>(AssocStmt->getCapturedStmt())) {
+      AssociatedStmt = InnerCapturedStmt;
+    }
```
```suggestion
  if (auto *AssocStmt = dyn_cast<CapturedStmt>(AssociatedStmt)) {
    if (auto *InnerCapturedStmt =
            dyn_cast<CapturedStmt>(AssocStmt->getCapturedStmt()))
      AssociatedStmt = InnerCapturedStmt;
  }
```

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -4529,6 +4529,191 @@ void CodeGenFunction::EmitOMPMasterDirective(const OMPMasterDirective &S) {
+      OMPLexicalScope Scope(*this, S, OMPD_dispatch);
+      emitIfElse(this, AssociatedStmt, NoVariantsCondition, NoContextCondition);
+    }
+  } else
+    EmitStmt(AssociatedStmt);
```
```suggestion
  } else {
    EmitStmt(AssociatedStmt);
  }
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -4280,6 +4282,15 @@ getTargetRegionParams(Sema &SemaRef) {
+  ASTContext &Context = SemaRef.getASTContext();
+  SmallVector<SemaOpenMP::CapturedParamNameType> Params;
+
+  Params.push_back(std::make_pair(StringRef(), QualType()));
```
```suggestion
  Params.emplace_back(StringRef(), QualType());
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -10554,6 +10687,8 @@ SemaOpenMP::ActOnOpenMPDispatchDirective(ArrayRef<OMPClause *> Clauses,
     return StmtError();
 
   Stmt *S = cast<CapturedStmt>(AStmt)->getCapturedStmt();
+  if (isa<CapturedStmt>(S))
+    S = cast<CapturedStmt>(S)->getCapturedStmt();
```
```suggestion
  if (auto *CS = dyn_cast<CapturedStmt>(S))
    S = CS->getCapturedStmt();
```

alexey-bataev commented on clang/test/OpenMP/dispatch_codegen.cpp:
```diff
@@ -0,0 +1,364 @@
+// expected-no-diagnostics
```
The tests with classes and function members are required, as well as classes, used in conditions

Commit:
Minor changes based on review feedback.
